### PR TITLE
Mm 378 - GitHub - Check for an existing webhook when subscribing to a repo

### PR DESCRIPTION
**Pull request by sibasankar-demansol**
when the [owner / repo] is subscribed , check if have setup any webhook if not let user know they need to do it

ticket here
https://github.com/mattermost/mattermost-plugin-github/issues/378

### CODE REVIEW

**Comment by mickmister** [comment]
@hanzei There's an issue here @sibasankar-demansol ran into where the OAuth token does not have permissions to create webhooks. We currently request the scopes:

- repo
- notifications
- read:org

Looking at the [available scopes](https://github.com/google/go-github/blob/0d84fe1b2f64d2813bb97e0fd5f1df4963a2c3a7/github/authorizations.go#L20-L44), we can add these for admin users:

- read:repo_hook
- write:repo_hook
- admin:repo_hook
- admin:org_hook

We can then have a `/github connect admin` command to apply these scopes. I also think we should be storing the scopes in the `GitHubUserInfo` struct, so we always know what permissions a user has inside of GitHub.

@hanzei Let me know your thoughts on this, and I'll make a separate ticket for the task.

File: server/plugin/command.go [hunk 1]
```diff
@@ -453,6 +469,151 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+		var config = make(map[string]interface{})
+
+		if len(parameters) > 2 {
+			for _, val := range strings.Split(parameters[2], ",") {
+				KVPair := strings.Split(val, "=")
+				switch {
+				case KVPair[0] == "insecure_ssl":
+					insecureSSL, err := strconv.ParseBool(KVPair[1])
+					if err != nil {
+						return "config filed insecure_ssl can only have on of the following true or false "
```
Comment (sanjaydemansol) [inline server/plugin/command.go:502]:
```suggestion
						return "Config failed, `insecure_ssl` must have a boolean(true|false) value."
```

File: server/plugin/command.go [hunk 2]
```diff
@@ -453,6 +469,151 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+				hookURL += "/" + repo
+			}
+			hookURL += "/settings/hooks/"
+			txt += fmt.Sprintf(" *  [%s](%s) :  -  %s", hookID, hookURL+hookID, strings.Join(hook.Events, " , "))
+			txt += newLine
+		}
+
+		return txt
+	}
+	return "invalid action only add and list command are supported"
```
Comment (sanjaydemansol) [inline server/plugin/command.go:597]:
```suggestion
	return "Invalid action, only `add` and `list` commands are supported."
```

File: server/plugin/template.go [hunk 3]
```diff
@@ -350,7 +350,57 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 		"  * `/github mute list` - list your muted GitHub users\n" +
 		"  * `/github mute add [username]` - add a GitHub user to your muted list\n" +
 		"  * `/github mute delete [username]` - remove a GitHub user from your muted list\n" +
-		"  * `/github mute delete-all` - unmute all GitHub users\n"))
+		"  * `/github mute delete-all` - unmute all GitHub users\n" +
+		"* `/github Webhook` - Managed to List Webhook of [owner/repo] and add webhook to owoner[/repo].\n" +
```
Comment (sanjaydemansol) [inline server/plugin/template.go:354]:
```suggestion
		"  * `/github Webhook` - Manage all webhooks.\n" +
```

File: server/plugin/command.go [hunk 4]
```diff
@@ -453,6 +469,151 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+					ShouldCallNext = false
+				}
+			}
+		}
+
+		var txt string
+		if len(githubHooks) == 0 {
+			txt = "Currently there are no webhook in this [Owner/Repo]"
+		} else {
+			txt = "### Webhook in this Repositories\n"
```
Comment (sanjaydemansol) [inline server/plugin/command.go:581]: could you add [Owner/Repo] details to the heading?

File: server/plugin/command.go [hunk 5]
```diff
@@ -307,14 +313,21 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
...
 	if repo == "" {
 		if err := p.SubscribeOrg(ctx, githubClient, args.UserId, owner, args.ChannelId, features, flags); err != nil {
 			return err.Error()
 		}
-
-		return fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+		msg := fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+		githubHooks, _, _ := githubClient.Repositories.ListHooks(ctx, owner, repo, &github.ListOptions{PerPage: page})
+		if len(githubHooks) == 0 {
+			msg += fmt.Sprintf("can configure the webhook to receive notification [here](%s).", webhookURL)
```
[the remark this answers is not in the document]
Comment (sanjaydemansol) [reply server/plugin/command.go:328]: Also, we can show something like **_contact admin_** if user is not Matermost Admin.

**Review by sanjaydemansol — COMMENTED** [review]
Suggestions

File: server/plugin/command.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (maisnamraju) [reply server/plugin/command.go:328]: I would rephrase it to "Configure webhook to receive notification [here] (%s)"

File: server/plugin/command.go [hunk 6]
```diff
@@ -453,6 +469,151 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	}
+	command := parameters[0]
+	parameters = parameters[1:]
+	ctx := context.Background()
+	githubClient := p.getGithubClient(userInfo)
+	baseURL := p.getBaseURL()
+	switch command {
+	case add:
+		if len(parameters) < 1 {
+			return "Unknown action, Currently supported to add and List webhook "
```
[the remark this answers is not in the document]
Comment (maisnamraju) [reply server/plugin/command.go:484]: lgtm

File: server/plugin/command.go [hunk 7]
```diff
@@ -453,6 +469,151 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+		}
+		owner, repo := parseOwnerAndRepo(parameters[0], p.getBaseURL())
+		if owner == "" {
+			return "Currently supported to add webhook to a specific [owner/repo] ."
```
[the remark this answers is not in the document]
Comment (maisnamraju) [reply server/plugin/command.go:488]: LGTM

File: server/plugin/command.go (hunk 2 above)
Comment (maisnamraju) [reply server/plugin/command.go:597]: LGTM

File: server/plugin/template.go (hunk 3 above)
Comment (maisnamraju) [reply server/plugin/template.go:354]: Just need to get rig of the white space before the `*` at 354

File: server/plugin/command.go [hunk 8]
```diff
@@ -615,6 +776,30 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
+
+	webhook.AddCommand(webhookList)
+	webhookAdd := model.NewAutocompleteData(add, "[owner/repo] callback_url token content_type events insecure_ssl", "Add a webhook to desired owner[/repo] with optional --config")
+	webhookAdd.AddTextArgument("Owner/repo to create a webhook", "owner[/repo]", "")
+	webhookAdd.AddTextArgument("Call Back URL", "", "")
+	webhookAdd.AddTextArgument("token", "", "")
+	webhookAdd.AddTextArgument("content_type", "", "")
+	webhookAdd.AddTextArgument("Comma-delimited list of one or more of: create, delete, check_run, check_suite, code_scanning_alert, member, commit_comment, deploy_key, deployment_status, deployment, discussion_comment, discussion, fork, issue_comment, issues, label, meta, milestone, package, page_build, project_card, project_column, project, pull_request_review_comment, pull_request_review_thread, pull_request_review, pull_request, push, registry_package, release, repository, repository_import, repository_vulnerability_alert, secret_scanning_alert, star, status, team_add, public, watch, gollum . Defaults to *", "[features] (optional)", `/[^,-\s]+(,[^,-\s]+)*/`)
```
Comment (mickmister) [inline server/plugin/command.go:789]: I think the point of this feature is to make it so the admin doesn't need to think about the webhook. We don't want to simply "move" the burden to our platform. Here's how I think this feature should work:

The admin runs `/github subscriptions add ...` to create a subscription. We check to see if we have a webhook set up for this repo or the surrounding org. If one exists, we're good to go and nothing left to do except create our own internal subscription as usual.

Otherwise, we reply back to the user: "No webhook was found for this repository or organization. Would you like the webhook to be created?", and show a form in the ephemeral response using an [Interactive Message](https://docs.mattermost.com/developer/interactive-messages.html). If the user has privileges to create an org-level webhook, provide that as an option on the form.

When the form is submitted, we create the appropriate webhook and automatically choose to the webhook events that the plugins needs to use. We can include the URL to the webhook that we created in our final output.

We should then process the original command submission of `/github subscriptions add ...`, so that information needs to be stored in the form. You can use the `Context` field of [PostActionIntegration](https://pkg.go.dev/github.com/mattermost/mattermost-server/model#PostActionIntegration
) to store the information about the originally submitted command.

Once that is done, we respond to the user with a link to the GitHub webhook page, and a message saying the subscription was created.

@sibasankar-demansol @hanzei @aaronrothschild Thoughts on this?

**Review by mickmister — CHANGES_REQUESTED** [review]

File: server/plugin/command.go (hunk 8 above)
Comment (hanzei) [reply server/plugin/command.go:789]: While an interactive message is a very handy way to create the webhook, I would also be fine simply making use of https://github.com/mattermost/mattermost-plugin-github/pull/456 and returning an ephemeral post that contains the command that needs to be run.

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on [this](https://github.com/mattermost/mattermost-plugin-github/pull/466#discussion_r685694061) discussion

File: server/plugin/command.go (hunk 8 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:789]: @hanzei  currently the scope of oAuth is for Repo level , could you let me know should we add that to scope 

rest suggestion i will work on

Comment (mickmister) [reply server/plugin/command.go:789]: @hanzei I like the slash command idea the ephemeral message idea. 2/5 I think we should show 2 commands, one for making a repo-level webhook, and one for making an org-level webhook.

Comment (sibasankar-demansol) [reply server/plugin/command.go:789]: @hanzei  , @mickmister  
i have few things from my side
1. currently #456  is yet to be done and is incomplete until we give the oAuth access for Org level or move on with creating webhook on Repo level .
2. Regarding this issue do you want me to show a ephemeral message with 2 commands for repo and org level differently , as the slash command is not yet completed to create webhook only way left is to give a link redirecting to github plugin console

Comment (mickmister) [reply server/plugin/command.go:789]: @sibasankarnayak 1/5 How about we assume the other PR is complete, and we put the two slash commands here. Then  we can put the "Do Not Merge/Awaiting PR" label on this PR to signal that it's waiting on changes.

Comment (sibasankarnayak) [reply server/plugin/command.go:789]: @mickmister  Thats sounds good,

just have a thought of show 
when user subscribe to a individual repo can show 2 commands
and when user subscribe to org then we can return one command to create org hook

what do you think does it makes any sense

Comment (sibasankarnayak) [reply server/plugin/command.go:789]: @mickmister  @hanzei 
can you suggest the changes over visual representation 
 Screenshot 2021-08-19 at 5 57 12 PM

Comment (hanzei) [reply server/plugin/command.go:789]: The response should make it clear that the command needs to be run. Also there is a  `/` missing at the beginning.

Are both webhooks needed in any case? Is the org specific enough?

Comment (sibasankarnayak) [reply server/plugin/command.go:810]: have updated the message , can you check now

Screenshot 2021-08-25 at 9 59 58 PM

Comment (mickmister) [reply server/plugin/command.go:810]: @sibasankarnayak Make sure you either mention or re-request review from the person you're communicating with on in issues and PRs. The other person doesn't get a push notification unless you do one or the other. The one time you don't need to mention someone is when you are commenting on their issue or PR. They will automatically get notifications for the comment.

When you have made changes that you need to be reviewed (which is the case here), your comment here is fine but you should also explicitly re-request review so the reviewer knows they need to review again.

Comment (mickmister) [reply server/plugin/command.go:810]: Actually since you wouldn't be requesting a full re-review here, you can just mention the reviewer in your comment cc @hanzei

Comment (sibasankarnayak) [reply server/plugin/command.go:914]: @mickmister  Sure will keep this in mind

File: server/plugin/command.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply server/plugin/command.go]: > Blocking on [this](https://github.com/mattermost/mattermost-plugin-github/pull/466#discussion_r685694061) discussion

Made the changes required

**Review by hanzei — CHANGES_REQUESTED** [review]
While testing this change I'm running into a similar issue as in https://github.com/mattermost/mattermost-plugin-github/pull/456. 
Screenshot from 2021-09-07 10-00-55

Is there anything specific requirement to test this PR?

File: plugin.json [hunk 9]
```diff
@@ -105,5 +105,50 @@
             }
         ],
         "footer": "* To report an issue, make a suggestion or a contribution, [check the repository](https://github.com/mattermost/mattermost-plugin-github)."
+    },
+    "props":{
```
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply plugin.json:109]: this is a branch created from issue  #379  , which is created for  adding variables to mainfest.go

**Comment by sibasankarnayak** [comment]
> While testing this change I'm running into a similar issue as in #456.
> Screenshot from 2021-09-07 10-00-55
> 
> Is there anything specific requirement to test this PR?

Can you let me know the Command you are using for ?

**Comment by hanzei** [comment]
@sibasankarnayak `/github hooks add matterpoll/matterpoll some_secret`

File: server/plugin/command.go [hunk 10]
```diff
@@ -506,6 +522,171 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	command := parameters[0]
+	parameters = parameters[1:]
+	ctx := context.Background()
+	githubClient := p.getGithubClient(userInfo)
+	baseURL := p.getBaseURL()
+	switch command {
+	case add:
+		if len(parameters) < 1 {
+			return "Invalid parameters for add command, Provide [owner/repo] details."
+		} else if len(parameters) < 2 {
```
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/command.go:538]:
```suggestion
		}
		
		if len(parameters) < 2 {
```

File: plugin.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin.json:109]: If the change isn't related to this PR, the change should be removed from this PR

[the remark this answers is not in the document]
Comment (mickmister) [reply plugin.json:109]: I haven't ever seen this pattern either. This is to share values between the server and webapp portions of the plugin?

File: server/plugin/command.go [hunk 11]
```diff
@@ -345,6 +349,12 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
 			}
 			subOrgMsg += "\n\n" + fmt.Sprintf("Notifications are disabled for %s", excludeMsg)
 		}
+		githubHooks, _, _ := githubClient.Repositories.ListHooks(ctx, owner, repo, &github.ListOptions{PerPage: page})
+		if len(githubHooks) == 0 {
+			subOrgMsg += "\n No webhook was found for this repository or organization. Would you like the webhook to be created?"
```
Comment (mickmister) [inline server/plugin/command.go:354]: I don't think we want the spaces after each `\n`. Can you go through the PR and remove the extra space at the beginning of each new line?

```suggestion
			subOrgMsg += "\nNo webhook was found for this repository or organization. Would you like the webhook to be created?"
```

File: plugin.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin.json:109]: To minimize the diff between this PR and the other PR that is in progress, you may be able to have this PR's base branch set to the other PR's branch. Though it may not let you do this because it's on a fork

**Review by mickmister — CHANGES_REQUESTED** [review]

File: plugin.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply plugin.json:109]: yes have tried creating base branch as the other one but it only let us to create PR with MM branch

**Comment by sibasankarnayak** [comment]
> matterpoll



> @sibasankarnayak `/github hooks add matterpoll/matterpoll some_secret`

@hanzei  seems the org/repo doesn't exist , this 404 error occurs only in 2 case 
1.  trying to create out of scope 
2.  passed invalid org/repo 

you need to pass a valid org/repo  i.e. `/github webhook add sibasankarnayak/mattermost-plugin-github some_secret htpps://callback_url`

File: server/plugin/command.go (hunk 11 above)
Comment (sibasankarnayak) [reply server/plugin/command.go:354]: Done

File: server/plugin/command.go [hunk 12]
```diff
@@ -345,6 +349,12 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
 			}
 			subOrgMsg += "\n\n" + fmt.Sprintf("Notifications are disabled for %s", excludeMsg)
 		}
+		githubHooks, _, _ := githubClient.Repositories.ListHooks(ctx, owner, repo, &github.ListOptions{PerPage: page})
```
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply server/plugin/command.go:352]: @hanzei i thing there is no need of pagination here , as the main objective is to find do we have any existing web hook , 

even fetching hooks for page 1 can confirm if any hook is already there or not
feel if the page 1 have 0 hooks there won't be any page 2

File: plugin.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply plugin.json:109]: > I haven't ever seen this pattern either. This is to share values between the server and webapp portions of the plugin?

yes in the #379  this was a suggestion to use by @hanzei  [here](https://github.com/mattermost/mattermost-plugin-github/pull/456#discussion_r670259213) 
after figuring out is there any way , i felt props is the only where we can add new values as rest is with static structure

File: server/plugin/command.go [hunk 13]
```diff
@@ -506,6 +522,171 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	return nil
+}
+func (p *Plugin) GetHook(ctx context.Context, githubClient *github.Client, owner, repo string, hookID int64) (*github.Hook, *github.Response, error) {
+	if repo != "" {
+		return githubClient.Repositories.GetHook(ctx, owner, repo, hookID)
+	}
+	return githubClient.Organizations.GetHook(ctx, owner, hookID)
+}
+func (p *Plugin) CreateHook(ctx context.Context, githubClient *github.Client, owner, repo string, hook github.Hook) (*github.Hook, *github.Response, error) {
+	p.API.LogWarn("log ", "githubClient", githubClient, "owner", owner)
```
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply server/plugin/command.go]: removed

File: server/plugin/command.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/command.go:348]: Does this call also list org-level webhooks? Also, how do we know the webhook is pointing to the Mattermost server, and that it has the correct settings?

File: server/plugin/command.go [hunk 14]
```diff
@@ -358,7 +369,13 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
 	} else if ghRepo != nil && ghRepo.GetPrivate() {
 		msg += "\n\n**Warning:** You subscribed to a private repository. Anyone with access to this channel will be able to read the events getting posted here."
 	}
-
+	githubHooks, _, _ := githubClient.Repositories.ListHooks(ctx, owner, repo, &github.ListOptions{PerPage: page})
+	if len(githubHooks) == 0 {
+		msg += "\nNo webhook was found for this repository or organization. Would you like the webhook to be created?"
+		msg += "\nSuggested Command, Can run any one command to create a wehook"
+		msg += fmt.Sprintf("\n`/github webhook add %s [secret:replace with webhook secret from Plugin console] [content-type:application-json/application/x-www-form-urlencoded] [features:create/delete..]  [ssecure-ssl:true/false]`", owner)
+		msg += fmt.Sprintf("\n`/github webhook add %s/%s [secret:replace with webhook secret from Plugin console] [content-type:application-json/application/x-www-form-urlencoded] [features:create/delete..]  [ssecure-ssl:true/false]`", owner, repo)
```
Comment (mickmister) [inline server/plugin/command.go:377]: @hanzei Should we put the secret here in the output, or have the admin go to the system console as noted here?

File: plugin.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin.json:109]: > yes have tried creating base branch as the other one but it only let us to create PR with MM branch

@sibasankarnayak Are you able to create branches in Mattermost's repo with your current permissions? cc @hanzei

[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply plugin.json:109]: @mickmister No tried creating a branch and push but unable to do so
cc @hanzei 
Screenshot 2021-09-10 at 7 48 02 AM

File: server/plugin/command.go (hunk 14 above)
Comment (sibasankarnayak) [reply server/plugin/command.go]: @mickmister @hanzei 
made a change where the secret is picked directly from github plugin console , no need to pass it explicitly

```
   /github webhook add sibasankarnayak/mattermost-server  callback_url
```

Note : callback url is only need to pass it runs on localhost as github doesn't permit to add localhost url

**Comment by mickmister** [comment]
@sanjaydemansol @maisnamrajusingh For any requests you've made on this PR that have been fulfilled, please click "Resolve Conversation" at the bottom of the request's discussion

File: server/plugin/command.go [hunk 15]
```diff
@@ -506,6 +522,171 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
 	}
 }
 
+func (p *Plugin) handlewebhook(_ *plugin.Context, args *model.CommandArgs, parameters []string, userInfo *GitHubUserInfo) string {
+	if len(parameters) == 0 {
+		return "Invalid webhook command. Available command are `add` and `list` ."
```
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/command.go]:
```suggestion
Invalid webhook command. Available commands are `add` and `list`.
```

File: server/plugin/command.go (hunk 14 above)
Comment (mickmister) [reply server/plugin/command.go]: If the site URL is configured to use `localhost`, can we just return an error saying "please set your site URL to something other than localhost". The secret and callback URL will then be optional parameters, then we can keep the command nice and short.

Comment (sibasankarnayak) [reply server/plugin/command.go]: @mickmister currently site url will give  error if it is local host and secret is not at all needed  and it is optional
picking up from plugin console directly

File: server/plugin/command.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply server/plugin/command.go:348]: Yes it can , but it can not verify automatically , 
Ul will be showing all the web-hook id which is hyperlink can visit and verify if they want

File: server/plugin/command.go (hunk 14 above)
Comment (mickmister) [reply server/plugin/command.go]: @sibasankarnayak Can we make it so the `callback_url` is optional? I was suggesting we assume that the site URL is not localhost in the main case, so the admin would not need to supply a `callback_url`.

The only time they need to supply a `callback_url` is if the site URL _is_ localhost. So let's not make them supply the `callback_url` when the site URL is not localhost.

Comment (sibasankarnayak) [reply server/plugin/command.go]: @mickmister current the callback url is optional , you can try it if your ulr is not local host

**Comment by maisnamrajusingh** [comment]
@mickmister doing now. will update you once i sync with @sibasankarnayak  on this.

**Comment by mickmister** [comment]
@hanzei I'm seeing this feature as an opportunity to support multiple webhook secrets. There was recently a request for this functionality. https://community-daily.mattermost.com/core/pl/uf5fzfmu97bbfqfw1qgjjebh4h

Out of scope of this PR, but if we move secrets to the kv store instead of plugin settings, any webhooks we automate the creation of we can store a separate secret for.

@sibasankarnayak Something else that comes to mind: Are we able to put some sort of "id" on the webhook request, so we can know if this plugin was the creator of the webhook? Then when we fetch the webhooks, we can definitively know which ones are for our Mattermost instance.

File: server/plugin/command.go [hunk 16]
```diff
@@ -501,6 +518,163 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+				if githubResponse.NextPage == 0 {
+					break
+				}
+				opt.Page = githubResponse.NextPage
+			}
+		}
+
+		var txt string
+		if len(githubHooks) == 0 {
+			txt = "Currently there are no webhook in this [Owner/Repo]"
```
Comment (mickmister) [inline server/plugin/command.go:635]: Can we be more specific here? Like:

```suggestion
			txt = "There are currently no GitHub webhooks created for the [%s](https://github.com/%s) org or the [%s/%s](https://github.com/%s/%s) repository."
```

Which should ultimately render as:

There are currently no GitHub webhooks created for the [mattermost](https://github.com/mattermost) org or the [mattermost/mattermost-plugin-github](https://github.com/mattermost/mattermost-plugin-github) repository.

File: server/plugin/command.go (hunk 4 above)
Comment (mickmister) [reply server/plugin/command.go:637]:
```suggestion
			txt = "### Webhooks in this Repository\n"
```

File: server/plugin/command.go [hunk 17]
```diff
@@ -501,6 +518,163 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+			if repo != "" {
+				hookURL += "/" + repo
+			}
+			hookURL += "/settings/hooks/"
+			txt += fmt.Sprintf(" *  [%s](%s) :  -  %s \n", hookID, hookURL+hookID, strings.Join(hook.Events, " , "))
+		}
+
+		return txt
+	}
+	return "Invalid action, only `add` and `list` commands are supported."
```
Comment (mickmister) [inline server/plugin/command.go:652]: Can we split `handlewebhook` into a few different functions, so this is a little more readable? Something like:

```go
switch command {
	case add:
            txt, err := p.handleWebhookAdd(args)
            // log err if not nil
            return txt
        case list:
            txt, err := p.handleWebhookList(args)
            // log err if not nil
            return txt
        default:
            return "Invalid action, only `add` and `list` commands are supported."
```

func (p *Plugin) handlewebhook(_ *plugin.Context, args *model.CommandArgs, parameters []string, userInfo *GitHubUserInfo) string {

File: server/plugin/command.go [hunk 18]
```diff
@@ -501,6 +518,163 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+}
+func (p *Plugin) CheckOptionsValid(options []string) error {
+	for _, val := range options {
+		for _, item := range Manifest.Props["webhook_events"].([]interface{}) {
+			if item.(string) != val {
+				return errors.New(val + " is not a valid events to trigger webhook")
+			}
+		}
+	}
```
Comment (mickmister) [inline server/plugin/command.go:661]: What's the point of having these in the manifest versus in the code?

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks for the updates on this PR :+1: I have some requests and questions

File: server/plugin/command.go [hunk 19]
```diff
@@ -501,6 +518,163 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+					if strings.Contains(val, "*") && len(val) == 1 {
+						hook.Events = []string{"*"}
+						break
+					}
+					if err := p.CheckOptionsValid(strings.Split(val, ",")); err != nil {
+						return err.Error()
+					}
+					hook.Events = strings.Split(val, ",")
+				}
+			}
```
[the remark this answers is not in the document]
Comment (sibasankarnayak) [reply server/plugin/command.go:562]: @mickmister tried with slash command but it was too confusing for user i felt

currently we need 5 parameters from user
[org/repo]  - 
callBackUrl - optional ( default to MMURL +"plugin/github/plugin")
content-type - optional ( default x-www-form-urlencoded)
event - optional ( * -default , create ,delete .... ) 
insecureSSL -  optional ( true / false , default to false) - already implemented with flags

but while parsing instead of taking in sequence as there are lot of parameter so tried this way,
if needed can do otherway , please let me know your view

File: server/plugin/command.go (hunk 18 above)
Comment (sibasankarnayak) [reply server/plugin/command.go]: Sorry , i misunderstood the comment from hanzei https://github.com/mattermost/mattermost-plugin-github/pull/456#discussion_r670259213 , fixed this in this commit.

File: server/plugin/command.go (hunk 17 above)
Comment (sibasankarnayak) [reply server/plugin/command.go]: Sure , implemented this

File: server/plugin/command.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/command.go]: > tried with slash command but it was too confusing for user i felt

Confusing in what way?

> currently we need 5 parameters from user

Since 4 are optional, I would say we only need 1, but that is beside the point. 2/5 we should use flags instead of guessing what field they are filling out. The required one can be positional if that is possible, but all the others should be flag arguments.

File: server/plugin/command.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/command.go:392]: @hanzei Please see my question above:

> Are we intending to have the plugin be a tool for managing other webhooks not related to the plugin? It seems like we're straying away from the purpose of the feature.

What purpose does the webhook command serve other than just accepting the org/repo to subscribe to? I'm not sure I understand why we support so many arguments when the main use case is to configure the webhooks for this plugin, which only requires the admin to provide the org or repo.

File: server/plugin/command.go [hunk 20]
```diff
@@ -501,6 +562,173 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	}
+
+	hookURL := baseURL + owner
+	if repo != "" {
+		hookURL += "/" + repo
+	}
+	hookURL += "/settings/hooks/"
+	txt := "Webhook Created Successfully \n"
+	hookID := strconv.Itoa(int(*githubHook.ID))
+	txt += fmt.Sprintf(" *  [%s](%s) :  -  %s \n", hookID, hookURL+hookID, strings.Join(hookDetails.Events, " , "))
```
Comment (mickmister) [inline server/plugin/command.go:629]: 0/5 what do you think of this approach?

```suggestion
	txt := "Webhook Created Successfully \n"

	hookURL += "/settings/hooks"
	events := strings.Join(hookDetails.Events, " , ")
	txt += fmt.Sprintf(" *  [%s](%s/%v) :  -  %s \n", hookID, hookURL, *githubHook.ID, events)
```

File: server/plugin/command.go [hunk 21]
```diff
@@ -501,6 +562,173 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+		}
+		hookURL += "/settings/hooks/"
+		txt += fmt.Sprintf(" *  [%s](%s) :  -  %s \n", hookID, hookURL+hookID, strings.Join(hook.Events, " , "))
+	}
+
+	return txt
+}
+func (p *Plugin) handlewebhook(_ *plugin.Context, args *model.CommandArgs, parameters []string, userInfo *GitHubUserInfo) string {
+	if len(parameters) == 0 {
+		return "Invalid webhook command. Available command are `add` and `list` ."
```
Comment (mickmister) [inline server/plugin/command.go:693]:
```suggestion
		return "Invalid webhook command. Available commands are `add` and `list` ."
```

File: server/plugin/command.go [hunk 22]
```diff
@@ -501,6 +562,173 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	case list:
+		return p.handleWebhookList(ctx, parameters, baseURL, githubClient)
+	}
+	return "Invalid action, only `add` and `list` commands are supported."
+}
+func (p *Plugin) CheckOptionsValid(options []string) error {
+	for _, val := range options {
+		for _, item := range webhookEvents {
+			if item != val {
+				return errors.New(val + " is not a valid events to trigger webhook")
```
Comment (mickmister) [inline server/plugin/command.go:712]:
```suggestion
				return errors.New(val + " is not a valid event to trigger webhook")
```

File: server/plugin/command.go [hunk 23]
```diff
@@ -501,6 +562,173 @@ func (p *Plugin) handleIssue(_ *plugin.Context, args *model.CommandArgs, paramet
...
+	baseURL := p.getBaseURL()
+	switch command {
+	case add:
+		return p.handleWebhookAdd(ctx, parameters, baseURL, githubClient)
+	case list:
+		return p.handleWebhookList(ctx, parameters, baseURL, githubClient)
+	}
+	return "Invalid action, only `add` and `list` commands are supported."
+}
+func (p *Plugin) CheckOptionsValid(options []string) error {
```
Comment (mickmister) [inline server/plugin/command.go:708]: Functions should always have a space between them. Not sure how CI didn't catch this

```suggestion
}

func (p *Plugin) CheckOptionsValid(options []string) error {
```

File: server/plugin/command.go [hunk 24]
```diff
@@ -340,6 +389,12 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
 			}
 			subOrgMsg += "\n\n" + fmt.Sprintf("Notifications are disabled for %s", excludeMsg)
 		}
+		githubHooks, _, _ := githubClient.Repositories.ListHooks(ctx, owner, repo, &github.ListOptions{PerPage: page})
+		if len(githubHooks) == 0 {
+			subOrgMsg += "\nNo webhook was found for this repository or organization. Would you like the webhook to be created?"
+			subOrgMsg += "\nSuggested Command to create a wehook"
+			subOrgMsg += "\n`/github webhook add " + owner + "[content-type:application-json/application/x-www-form-urlencoded] [features:create/delete..]  [ssecure-ssl:true/false]`"
```
Comment (mickmister) [inline server/plugin/command.go:396]: We should only be showing the standard command. Also, please put this text in a shared function to be used in the other case where we output this information.

File: server/plugin/command.go (hunk 8 above)
Comment (mickmister) [reply server/plugin/command.go:955]: > The response should make it clear that the command needs to be run. Also there is a `/` missing at the beginning.
> 
> Are both webhooks needed in any case? Is the org specific enough? Can the text make the answer to these questions clear?

@sibasankarnayak Please make sure you address @hanzei's points here

**Comment by mickmister** [comment]
There are a lot of unresolved comments on this PR. @sanjaydemansol @maisnamrajusingh Can you please resolve the conversations for any requests you've made on this PR that have been fulfilled? Please click "Resolve Conversation" at the bottom of the request's discussion.

**Comment by dipak-demansol** [comment]
@mickmister some days ago this PR assigned directly to me by siba but as per your recent comment can you pls tell me should i test this PR?

**Review by hanzei — COMMENTED** [review]
Let's keep this PR on the side until https://github.com/mattermost/mattermost-plugin-github/pull/456 is merged as they will heavily conflict

File: server/plugin/command.go (hunk 8 above)
Comment (mickmister) [reply server/plugin/command.go]: @sibasankarnayak We should make it clear that one command will create a webhook for the repo, and the other will create one for the whole org. And we should explain that if the webhook is made for the org, then you won't need to create another webhook for any repos in that org going forward.
